Add reflection-based Search index definition

idg10 opened the pull request:
This implements https://github.com/Azure/azure-sdk-for-net/issues/2427

For the unsquashed history see https://github.com/idg10/azure-sdk-for-net/tree/reflection-schema which includes discussion with @brjohnstmsft as he reviewed my work on this feature.

idg10 committed:
Add reflection-based Search index definition

azuresdkci commented:
Can one of the admins verify this patch?

azurecla commented:
Hi __@idg10__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

azurecla commented:
__@idg10__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, AZPRBOT;

brjohnstmsft commented on src/Search/Microsoft.Azure.Search/Customizations/Indexes/FieldBuilder.cs:
```diff
@@ -0,0 +1,176 @@
+                    }
+                    else if ((indexAnalyzerAttribute = attribute as IndexAnalyzerAttribute) != null)
+                    {
+                        field.IndexAnalyzer = AnalyzerName.Create(((IndexAnalyzerAttribute) attribute).Name);
```
I missed this earlier. It looks like the cast to `IndexAnalyzerAttribute` is unnecessary if you use the `indexAnalyzerAttribute` variable instead.

idg10 replied:
Fixed in https://github.com/Azure/azure-sdk-for-net/pull/2536/commits/f46c44cde80ec93d73b15c292ff31f1ca72d531c (resquashed)

brjohnstmsft commented on src/Search/Microsoft.Azure.Search/Customizations/Indexes/Models/AnalyzerName.cs:
```diff
@@ -466,31 +466,515 @@ public sealed class AnalyzerName : ExtensibleEnum<AnalyzerName>
+            /// Standard ASCII Folding Lucene analyzer.
+            /// <see href="https://msdn.microsoft.com/library/azure/mt605304.aspx#Analyzers" /> 
+            /// </summary>
+            public static readonly AnalyzerName StandardAsciiFoldingLucene =
```
This should be a string.

idg10 replied:
Fixed in https://github.com/Azure/azure-sdk-for-net/pull/2536/commits/f46c44cde80ec93d73b15c292ff31f1ca72d531c (resquashed)

idg10 commented:
I've fixed both of the problems you found (the unnecessary repeated cast, and my bizarre error in `AnalyzerName.AsString.StandardAsciiFoldingLucene`). I presumed you still want this to be a single commit in the event that you accept the PR, so I've squashed it again and force pushed to my PR branch.

markcowl commented:
@azuresdkci add to whitelist

markcowl commented:
LGT^M once the build passes

brjohnstmsft commented:
@markcowl The build passed; Please merge
